springfox-core/src/main/java/springfox/documentation/schema/Model.java

```diff
@@ -96,8 +107,42 @@ public String getDiscriminator() {
   public ResolvedType getType() {
     return type;
   }
+  
+  public Boolean isMap() {
+      return isMapType;
+    }
 
   public String getExample() {
     return example;
   }
+  
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(id, name, type, qualifiedType, isMapType, properties, description, baseModel, discriminator, subTypes, example);
+  }
+  
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    Model that = (Model) o;
+
+    return Objects.equal(id, that.id) &&
+        Objects.equal(name, that.name) &&
+        Objects.equal(type, that.type) &&
+        Objects.equal(isMapType, that.isMapType) &&
+        Objects.equal(qualifiedType, that.qualifiedType) &&
+        Objects.equal(properties, that.properties) &&
+        Objects.equal(description, that.description) &&
+        Objects.equal(baseModel, that.baseModel) &&
+        Objects.equal(discriminator, that.discriminator) &&
+        Objects.equal(subTypes, that.subTypes) &&
+        Objects.equal(example, that.example);
+  }
```

Comment:
Dilip, I'm sorry, I didn't catch about Model related objects... For now, _Model_, _ModelRef_, _ResolvedMethodParameter_, _ModelProperty_, _AllowableRangeValues_ and _AllowableListValues_ are **immutable**.  Model map is a HashMap<String, Model>, where string's hash is being used for storing models.

 For caching _ModelContext_ class is used and ModelProvider returns List (not Set) with ModelContexts tree. (dependencies)